### PR TITLE
Third-party

### PR DESCRIPTION
**Pull request by dominiccaron**
Modifications to the management of third-party files by the project and by the build system, namely:
- relocate third-party source files to an appropriate subdirectory of ```ThirdParty```;
- create README within ```ThirdParty``` with disclaimer, information on and licenses of the original projects;
- update manner in which CMake compiles third-party files.

In reference to the last point, third-party files now are compiled into an auxiliary library that serves as a dependency for programs that link to the principal ChiTech library, including the main ChiTech executable itself. Modifications have been made to the CMake macros that are distributed with the project so that users linking their programs to the ChiTech library following the approach described in the documentation (```ChiDoc/PAGES/ProgrammersManual/devman_01_02_usinglib.h```) are unaffected by these changes.

Finally, with reference to pull request #168, these modifications allow to suppress warnings arising from third-party code through:
- the treatment of third-party headers as system headers;
- the application of a less stringent set of compiler flags to those files that are not under the direct management of the project.

### CODE REVIEW

File: ThirdParty/paraview/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,10 @@
+file(GLOB_RECURSE SOURCES "${CMAKE_CURRENT_SOURCE_DIR}/*.cxx")
+
+set(TARGET ThirdPartyParaView)
+add_library(${TARGET} OBJECT ${SOURCES})
+if (VTK_VERSION VERSION_LESS "8.90.0")
+    target_include_directories(${TARGET} SYSTEM ${VTK_INCLUDE_DIRS})
```
Comment (Naktakala) [inline ThirdParty/paraview/CMakeLists.txt]: This line gives me cmake error when I use VTK8.2. 
```
CMake Error at ThirdParty/paraview/CMakeLists.txt:6 (target_include_directories):
  target_include_directories called with invalid arguments
```
If I change it to just
```cmake
include_directories(SYSTEM ${VTK_INCLUDE_DIRS})
```
it seems to work just fine.

**Review by Naktakala — CHANGES_REQUESTED** [review]

**Review by zachhardy — COMMENTED** [review]
While the warnings from before did go away, this will not successfully compile on my machine (Apple M1 with GCC-11.1.0 and VTK-9.0.1) as is, while the development branch does. The error is below:

```
Undefined symbols for architecture arm64:
  "__ZN13vtkObjectBase11PrintHeaderERSo9vtkIndent", referenced from:
      __ZTV24vtkCleanUnstructuredGrid in libThirdParty.a(vtkCleanUnstructuredGrid.cxx.o)
  "__ZN13vtkObjectBase12PrintTrailerERSo9vtkIndent", referenced from:
      __ZTV24vtkCleanUnstructuredGrid in libThirdParty.a(vtkCleanUnstructuredGrid.cxx.o)
  "__ZN28vtkUnstructuredGridAlgorithm9PrintSelfERSo9vtkIndent", referenced from:
      __ZN24vtkCleanUnstructuredGrid9PrintSelfERSo9vtkIndent in libThirdParty.a(vtkCleanUnstructuredGrid.cxx.o)
ld: symbol(s) not found for architecture arm64
collect2: error: ld returned 1 exit status
make[3]: *** [../bin/ChiTech] Error 1
make[2]: *** [CMakeFiles/ChiTech.dir/all] Error 2
make[1]: *** [all] Error 2
make: *** [subsystem] Error 2
```

File: ThirdParty/paraview/CMakeLists.txt (hunk 1 above)
Comment (dominiccaron) [reply ThirdParty/paraview/CMakeLists.txt]: Please try again with the revision.

**Comment by dominiccaron** [comment]
> While the warnings from before did go away, this will not successfully compile on my machine (Apple M1 with GCC-11.1.0 and VTK-9.0.1) as is, while the development branch does. The error is below:
> 
> ```
> Undefined symbols for architecture arm64:
>   "__ZN13vtkObjectBase11PrintHeaderERSo9vtkIndent", referenced from:
>       __ZTV24vtkCleanUnstructuredGrid in libThirdParty.a(vtkCleanUnstructuredGrid.cxx.o)
>   "__ZN13vtkObjectBase12PrintTrailerERSo9vtkIndent", referenced from:
>       __ZTV24vtkCleanUnstructuredGrid in libThirdParty.a(vtkCleanUnstructuredGrid.cxx.o)
>   "__ZN28vtkUnstructuredGridAlgorithm9PrintSelfERSo9vtkIndent", referenced from:
>       __ZN24vtkCleanUnstructuredGrid9PrintSelfERSo9vtkIndent in libThirdParty.a(vtkCleanUnstructuredGrid.cxx.o)
> ld: symbol(s) not found for architecture arm64
> collect2: error: ld returned 1 exit status
> make[3]: *** [../bin/ChiTech] Error 1
> make[2]: *** [CMakeFiles/ChiTech.dir/all] Error 2
> make[1]: *** [all] Error 2
> make: *** [subsystem] Error 2
> ```

@zachhardy Do you confirm that you are working in a clean build environment? It should be sufficient to re-run cmake on the project; however, if that continues to give problems, you can attempt also to first remove the `chi_build` directory and then re-run cmake. The error message that you show identifies three methods that are updated in the move to the most recent version of `vtkCleanUnstructuredGrid`. Let us confirm that there are no compile artefacts that contribute to the issue before investigating other possible causes. Let me know.

**Comment by zachhardy** [comment]
> > While the warnings from before did go away, this will not successfully compile on my machine (Apple M1 with GCC-11.1.0 and VTK-9.0.1) as is, while the development branch does. The error is below:
> > ```
> > Undefined symbols for architecture arm64:
> >   "__ZN13vtkObjectBase11PrintHeaderERSo9vtkIndent", referenced from:
> >       __ZTV24vtkCleanUnstructuredGrid in libThirdParty.a(vtkCleanUnstructuredGrid.cxx.o)
> >   "__ZN13vtkObjectBase12PrintTrailerERSo9vtkIndent", referenced from:
> >       __ZTV24vtkCleanUnstructuredGrid in libThirdParty.a(vtkCleanUnstructuredGrid.cxx.o)
> >   "__ZN28vtkUnstructuredGridAlgorithm9PrintSelfERSo9vtkIndent", referenced from:
> >       __ZN24vtkCleanUnstructuredGrid9PrintSelfERSo9vtkIndent in libThirdParty.a(vtkCleanUnstructuredGrid.cxx.o)
> > ld: symbol(s) not found for architecture arm64
> > collect2: error: ld returned 1 exit status
> > make[3]: *** [../bin/ChiTech] Error 1
> > make[2]: *** [CMakeFiles/ChiTech.dir/all] Error 2
> > make[1]: *** [all] Error 2
> > make: *** [subsystem] Error 2
> > ```
> 
> @zachhardy Do you confirm that you are working in a clean build environment? It should be sufficient to re-run cmake on the project; however, if that continues to give problems, you can attempt also to first remove the `chi_build` directory and then re-run cmake. The error message that you show identifies three methods that are updated in the move to the most recent version of `vtkCleanUnstructuredGrid`. Let us confirm that there are no compile artefacts that contribute to the issue before investigating other possible causes. Let me know.

Yes it was a clean build environment. I just retested and the problem persisted.

**Comment by dominiccaron** [comment]
> > > While the warnings from before did go away, this will not successfully compile on my machine (Apple M1 with GCC-11.1.0 and VTK-9.0.1) as is, while the development branch does. The error is below:
> > > ```
> > > Undefined symbols for architecture arm64:
> > >   "__ZN13vtkObjectBase11PrintHeaderERSo9vtkIndent", referenced from:
> > >       __ZTV24vtkCleanUnstructuredGrid in libThirdParty.a(vtkCleanUnstructuredGrid.cxx.o)
> > >   "__ZN13vtkObjectBase12PrintTrailerERSo9vtkIndent", referenced from:
> > >       __ZTV24vtkCleanUnstructuredGrid in libThirdParty.a(vtkCleanUnstructuredGrid.cxx.o)
> > >   "__ZN28vtkUnstructuredGridAlgorithm9PrintSelfERSo9vtkIndent", referenced from:
> > >       __ZN24vtkCleanUnstructuredGrid9PrintSelfERSo9vtkIndent in libThirdParty.a(vtkCleanUnstructuredGrid.cxx.o)
> > > ld: symbol(s) not found for architecture arm64
> > > collect2: error: ld returned 1 exit status
> > > make[3]: *** [../bin/ChiTech] Error 1
> > > make[2]: *** [CMakeFiles/ChiTech.dir/all] Error 2
> > > make[1]: *** [all] Error 2
> > > make: *** [subsystem] Error 2
> > > ```
> > 
> > 
> > @zachhardy Do you confirm that you are working in a clean build environment? It should be sufficient to re-run cmake on the project; however, if that continues to give problems, you can attempt also to first remove the `chi_build` directory and then re-run cmake. The error message that you show identifies three methods that are updated in the move to the most recent version of `vtkCleanUnstructuredGrid`. Let us confirm that there are no compile artefacts that contribute to the issue before investigating other possible causes. Let me know.
> 
> Yes it was a clean build environment. I just retested and the problem persisted.

I will follow-up with you offline in order to obtain additional details.

**Comment by dominiccaron** [comment]
After additional investigation, it appears that compilation and/or execution of ParaView on Apple devices with the M1 chip is not trivial and, at present, is not supported by ParaView. The behaviour mentioned by @zachhardy bears similarities to errors reported elsewhere in relation to this issue. Although it would be nice to investigate further in order to identify a possible clean, configuration-level solution to the problem, my effectiveness is limited by the lack of personal access to the hardware in question. Consequently, a workaround has been developed (commit 2a0aa6e36d36e18f0d5c8bbc1774ff6730a52e11) that involves direct modification of the source code and impacts only some print functions. According to the ParaView license, redistribution of modified source files is permitted. We can decide whether or not to open an issue to remind ourselves to periodically monitor the development status of ParaView in an attempt to restore the ParaView files to their original form.

**Review by zachhardy — APPROVED** [review]

**Review by Naktakala — APPROVED** [review]

**Pull request merged**